File: app/src/main/java/inc/flide/eightvim/keyboardHelpers/KeyboardAction.java
package inc.flide.eightvim.keyboardHelpers;

/**
 * Created by flide on 23/11/15.
 */
public class KeyboardAction {
    public enum KeyboardActionType {INPUT_TEXT, INPUT_KEY, INPUT_SPECIAL};

    private KeyboardActionType keyboardActionType;
    private String text;
    private String capsLockText;
    private int keyEventCode;

    public KeyboardAction(KeyboardActionType keyboardActionType, String text, String capsLockText, int keyEventCode) {
        this.keyboardActionType = keyboardActionType;
        this.text = text;
        this.keyEventCode = keyEventCode;
        setCapsLockText(capsLockText);
    }

    private void setCapsLockText(String capsLockText) {
        if((capsLockText==null || capsLockText.length()==0) && this.text!=null){
            this.capsLockText = this.text.toUpperCase();
        } else{
            this.capsLockText = capsLockText;
        }
    }

    public String getText() {
        return text;
    }

    public int getKeyEventCode() {
        return keyEventCode;
    }

    public KeyboardActionType getKeyboardActionType() {
        return keyboardActionType;
    }

    public String getCapsLockText() { return capsLockText; }
}

File: 8vim/src/main/kotlin/inc/flide/vim8/ime/layout/models/KeyboardAction.kt
package inc.flide.vim8.ime.layout.models

import android.content.Context
import android.view.KeyEvent
import arrow.optics.optics
import inc.flide.vim8.lib.ExcludeFromJacocoGeneratedReport

@ExcludeFromJacocoGeneratedReport
@optics
data class KeyboardAction(
    val keyboardActionType: KeyboardActionType,
    val text: String,
    val capsLockText: String = "",
    val keyEventCode: Int,
    val keyFlags: Int,
    val layer: LayerLevel
) {
    companion object {
        val UNSPECIFIED = KeyboardAction(
            keyboardActionType = KeyboardActionType.INPUT_KEY,
            text = "",
            keyEventCode = 0,
            keyFlags = 0,
            layer = LayerLevel.FIRST
        )
    }
}

fun KeyboardAction.name(context: Context): String {
    return if (keyboardActionType == KeyboardActionType.INPUT_TEXT) {
        val sb = StringBuilder()
        if (text.isNotEmpty()) sb.append("""Lower case: "$text"""")
        if (text.isNotEmpty() && capsLockText.isNotEmpty()) sb.append("/")
        if (capsLockText.isNotEmpty()) sb.append("""Upper case: "$capsLockText"""")
        sb.toString()
    } else {
        if (keyEventCode < 0) {
            CustomKeycode.fromInt(keyEventCode).name(context)
        } else {
            KeyEvent.keyCodeToString(keyEventCode)
        }
    }
}
